From: Nick Rosbrook <enr0n@ubuntu.com>
Date: Thu, 18 Jan 2024 11:25:58 -0500
Subject: test: skip test-execute in arhmf LXC containers

This has the effect of skipping test-execute on rmhhf autopkgtest. Many
of the failures we see in LXC are related to AppArmor and kernel
differences, and currently the host for armhf autopkgtest is Focal
running Linux 5.4. Thus, we may get quite different results in
autopkgtest infrastructure than if we ran (a) native armhf host, or (b)
armhf container on a newer kernel (like 6.6 in Noble).

We get better LXC coverage by tests-in-lxd on other arches, and we have
a history already of just skipping things in test-execute on LXC
wholesale when things fail on armhf. Instead of continuing that, just
skip test-execute entirely on armhf in LXC.

Forwarded: no, Ubuntu specific
---
 src/test/test-execute.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/test/test-execute.c b/src/test/test-execute.c
index 4a4f0f0..e062625 100644
--- a/src/test/test-execute.c
+++ b/src/test/test-execute.c
@@ -7,6 +7,7 @@
 
 #include "sd-event.h"
 
+#include "architecture.h"
 #include "capability-util.h"
 #include "cpu-set-util.h"
 #include "copy.h"
@@ -1441,6 +1442,9 @@ TEST(run_tests_unprivileged) {
 }
 
 static int intro(void) {
+        if (detect_container() == VIRTUALIZATION_LXC && uname_architecture() == ARCHITECTURE_ARM)
+                return log_tests_skipped("Appear to be running on armhf container");
+
 #if HAS_FEATURE_ADDRESS_SANITIZER
         if (strstr_ptr(ci_environment(), "travis") || strstr_ptr(ci_environment(), "github-actions"))
                 return log_tests_skipped("Running on Travis CI/GH Actions under ASan, see https://github.com/systemd/systemd/issues/10696");
